op-program/builder/main.go

```diff
@@ -0,0 +1,211 @@
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	proofFmt := filepath.Join(root, "op-program", "bin", "%d"+suffix+".json")
+	cmd = exec.Command(cannonBin, "run", "--proof-at", "=0", "--stop-at", "=1", "--input", prestate, "--meta", meta, "--proof-fmt", proofFmt, "--output", "")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	proof := filepath.Join(root, "op-program", "bin", "0"+suffix+".json")
+	target := filepath.Join(root, "op-program", "bin", "prestate-proof"+suffix+".json")
+	if err := os.Rename(proof, target); err != nil {
+		return fmt.Errorf("failed to rename proof file: %w", err)
+	}
+	log.Info("Prestate proof saved", "path", target)
+	return nil
+}
+
+func runBuildPrestates(cliCtx *cli.Context) error {
+	logger := setupLogger()
+	ctx := cliCtx.Context
+
+	releasesOnly := cliCtx.Bool("releases-only")
+
+	root, err := opservice.FindMonorepoRoot(".")
+	if err != nil {
+		return fmt.Errorf("failed to find monorepo root: %w", err)
+	}
+
+	type prestateInfo struct {
+		programELF string
+		version    versions.StateVersion
+		suffix     string
+	}
+	programELF := filepath.Join(root, "op-program", "bin", "op-program-client64.elf")
+	interopProgramELF := filepath.Join(root, "op-program", "bin", "op-program-client-interop.elf")
+
+	prestates := []prestateInfo{
+		{programELF, versions.GetCurrentVersion(), "-mt64"},
+		{interopProgramELF, versions.GetCurrentVersion(), "-interop"},
+	}
+	if !releasesOnly && versions.GetCurrentVersion() != versions.GetExperimentalVersion() {
+		prestates = append(prestates, prestateInfo{programELF, versions.GetExperimentalVersion(), "-mt64Next"})
+		prestates = append(prestates, prestateInfo{interopProgramELF, versions.GetExperimentalVersion(), "-interopNext"})
+	}
+	for _, prestate := range prestates {
+		logger.Info("Building prestate", "version", prestate.version, "suffix", prestate.suffix)
+		if err := buildPrestate(ctx, logger, prestate.programELF, prestate.version, prestate.suffix); err != nil {
+			return fmt.Errorf("failed to build dev %s prestate: %w", prestate.suffix, err)
+		}
+	}
+
+	if !releasesOnly {
+		// some tests expect a "next" prestate to exist. So let's fake them if they're not built.
+		copies := []struct {
+			src, dst string
+		}{
+			{filepath.Join(root, "op-program", "bin", "prestate-mt64.bin.gz"), filepath.Join(root, "op-program", "bin", "prestate-mt64Next.bin.gz")},
+			{filepath.Join(root, "op-program", "bin", "prestate-interop.bin.gz"), filepath.Join(root, "op-program", "bin", "prestate-interopNext.bin.gz")},
+		}
+		for _, copy := range copies {
+			data, err := os.ReadFile(copy.src)
+			if err != nil {
+				return fmt.Errorf("failed to read file %s: %w", copy.src, err)
+			}
+			if err := os.WriteFile(copy.dst, data, 0644); err != nil {
+				return fmt.Errorf("failed to copy %s: %w", copy.src, err)
+			}
+		}
```

Comment:
Shouldn't we only do this if we didn't build the next versions above (ie if `versions.GetCurrentVersion() == versions.GetExperimentalVersion()`? If I'm reading the current code right it would build the actual *Next prestates but then overwrite them here with the current release prestate.

Reply:
Indeed. The "next" prestates should be copied only if they weren't built. Will fix.